#' load_ldsc_gene_groups
#'
#' \code{load_ldsc_gene_groups} Loads in the GeneGroups files that are generated by the Create_LDSC_Annot function.
#' The GeneGroup files should all be within a folder. Each one contains the quantile groups into which a gene was 
#' placed for each celltype (with respect to a particular Single Cell Transcriptome dataset). The GeneGroup files are
#' expected to have names like:
#' 
#'      "GeneGroups_celltype_data_allKImouse_level1_thresh0_trim0_CutOff0_percentile_StepSize_0.1.csv"
#'      
#'          Where "celltype_data_allKImouse_level1_thresh0_trim0" is the name of the celltype_data file.
#' 
#' This function loads each GeneGroups and stores that data in a set of nested lists.
#'
#' @param ggPath Root folder for the GeneGroup files. Assumes that this folder contains a subfolder called 'ResFiles'. 
#' @param dropTerms A string passed as an argument to gsub which can be used to reduce the GeneGroup filenames to just
#' the base name for the celltype_data file. For instance, the following name:
#'     "GeneGroups_celltype_data_allKImouse_level1_thresh0_trim0_CutOff0_percentile_StepSize_0.1.csv"
#' Can be reduced to: "celltype_data_allKImouse_level1_thresh0_trim0" using "GeneGroups_|_CutOff0_percentile_StepSize_0.1.csv"
#' @return A list of lists containing the quantile group of each gene in the 'cell_groups' dataframe, and also the maximum 
#' proportion within each quantile group within 'max_prop_inGroup'.
#' @examples
#' geneGroupFolder="/Users/ns9/Desktop/LDSC Results/Phase3bRes/GeneGroups"
#' gene_groups = load_ldsc_gene_groups(geneGroupFolder)
#' @export
#load_ldsc_gene_groups <- function(ggPath,dropTerms="GeneGroups_|_CutOff0_percentile_StepSize_0.1.csv"){
load_ldsc_gene_groups <- function(ggPath){
    library("data.table")
    all_gene_groups = list.files(path=sprintf("%s",ggPath),recursive = TRUE)
    gene_groups = list()
    print(sprintf("Found %s gene group files",length(all_gene_groups)))
    count=0
    for(gg in all_gene_groups){
        count=count+1
        #fName = gsub("GeneGroups_|_CutOff0_percentile_StepSize_0.1.csv","",gg)
        fName = gsub("/.*","",gg)
        print(sprintf("Loading file %s of %s: %s",count,length(all_gene_groups),gg))
        #ggIN = read.csv(sprintf("%s/%s",ggPath,gg))
        ggIN = fread(sprintf("%s/%s",ggPath,gg),sep=",")
        ggIN$proportion=a.n(ggIN$proportion)
        gene_groups[[fName]]=list()
        for(cc in as.character(unique(ggIN$celltype))){
            ccIN = ggIN[ggIN$celltype==cc,]
            ccIN = ccIN[ccIN$percentile!=-1,]
            if(length(grep("hgnc",colnames(ggIN)))>0){ccIN = unique(ccIN[,c("hgnc_symbol","proportion","percentile")])}
            if(length(grep("mgi",colnames(ggIN)))>0){ccIN = unique(ccIN[,c("mgi_symbol","proportion","percentile")])}
            max_prop = rep(0,length(unique(ccIN$percentile)))
            names(max_prop) = sort(unique(ccIN$percentile))
            for(qq in sort(unique(ccIN$percentile))){
                ctQQ = ccIN[ccIN$percentile==qq,]
                max_prop[as.character(qq)]=max(ctQQ$proportion)
            }
            gene_groups[[fName]][[cc]]=list(cell_groups=ccIN,max_prop_inGroup=max_prop)
        }
    }
    return(gene_groups)
}